### PR TITLE
Improved Threshold Scan, Touch and other small fixes

### PR DESCRIPTION
**Pull request by krautech**
## Description

Brings `CARTOGRAPHER_THRESHOLD_SCAN` into line with `CARTOGRAPHER_TOUCH`

Based on KrauTech-Nightly with improvements.
Maintains same command params from previous iteration for backwards compatibility.

## Checklist

The following relevant macros have been tested:

- [x] `CARTOGRAPHER_CALIBRATE`
- [x] `PROBE`
- [x] `PROBE_ACCURACY`
- [x] `CARTOGRAPHER_THRESHOLD_SCAN`
- [x] `CARTOGRAPHER_TOUCH CALIBRATE=1`
- [x] `CARTOGRAPHER_TOUCH`
- [x] `CARTOGRAPHER_TOUCH METHOD=manual`
- [x] `BED_MESH_CALIBRATE`

### CODE REVIEW

File: scanner.py [hunk 1]
```diff
@@ -427,117 +510,591 @@ def start_touch(self, gcmd, touch_settings, verbose):
...
+                        )
+                        results.append(result)  # Append the result for later comparison
+                        # If all repeat attempts succeeded, save the threshold
+                        if consistent_results and override == 0:
+                            best_threshold = current_threshold
+                            self._save_threshold(best_threshold)
+                            gcmd.respond_info(
+                                f"Confirmed Optimal Threshold Found: {best_threshold}"
+                            )
+                            break
```
Comment (Jomik) [inline scanner.py:818]: Obsolete message. We also mention it on line 877.

File: scanner.py [hunk 2]
```diff
@@ -427,117 +510,591 @@ def start_touch(self, gcmd, touch_settings, verbose):
...
+                )
+            else:
+                std_dev = None
+                success = False
+
+            self.toolhead.wait_moves()
+            self.toolhead.flush_step_generation()
+            self.previous_probe_success = 1 if success else 0
+
+            # Return relevant data
```
Comment (Jomik) [inline scanner.py:1009]: Should we set `self.detect_threshold_z` to `original_threshold` again? We could do it in a `finally` clause. The entire function is wrapped in a `try`.

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scanner.py (hunk 2 above)
Comment (krautech) [reply scanner.py:1009]: L861

Comment (Jomik) [reply scanner.py:1009]: That is in a different function though?

Comment (krautech) [reply scanner.py:1009]: yes its in `cmd_SCANNER_THRESHOLD_SCAN` which loops `start_threshold_scan`

So once theyre finished it sets the `self.detect_threshold_z` back

Comment (Jomik) [reply scanner.py:1009]: Alrighty! I'll accept that 👍

File: scanner.py [hunk 3]
```diff
@@ -427,117 +510,619 @@ def start_touch(self, gcmd, touch_settings, verbose):
...
+            f"/tmp/scanner_touch_scan_{time.strftime('%Y%m%d_%H%M%S')}.csv"
+            if verbose == 1
+            else None
+        )
+
+        # Prepare to track results
+        results = []
+        has_increased_threshold_max = (
+            False  # Flag to track if threshold_max has been increased
+        )
```
[the remark this answers is not in the document]
Comment (krautech) [reply scanner.py]: Fixed

File: scanner.py [hunk 4]
```diff
@@ -427,117 +510,619 @@ def start_touch(self, gcmd, touch_settings, verbose):
...
+
+                if result["success"]:
+                    # Check if this result meets "good" criteria
+                    if result["retries"] <= max_acceptable_retries and (
+                        result["standard_deviation"] is not None
+                        and result["standard_deviation"] <= max_acceptable_std_dev
+                    ):
+                        # Increase threshold_max by 3 steps above the current threshold, only if it hasn't been increased before
+                        if not has_increased_threshold_max:
+                            threshold_max = current_threshold + 5 * step
```
[the remark this answers is not in the document]
Comment (krautech) [reply scanner.py]: done

File: scanner.py (hunk 1 above)
Comment (krautech) [reply scanner.py]: done :D

**Comment by krautech** [comment]
I'm not happy enough with this yet. It clearly has issues.

image

Either 1. Its not outputting text about the other Qualify attempts or 2. Its moving onto the next threshold instead of doing all 5 qualify attempts and checking the results.

File: scanner.py [hunk 5]
```diff
@@ -427,117 +516,594 @@ def start_touch(self, gcmd, touch_settings, verbose):
...
+                touch_settings = TouchSettings(
+                    initial_position,
+                    homing_position,
+                    vars["accel"],
+                    vars["speed"],
+                    vars["retract_dist"],
+                    vars["retract_speed"],
+                    vars["num_samples"],
+                    vars["tolerance"],
+                    confirmation_retries,
```
Comment (Jomik) [inline scanner.py:760]: Why do we use `confirmation_retries` here when we drop the result here, if it above `max_acceptable_retries` https://github.com/Cartographer3D/cartographer-klipper/blob/2dfe5f0158384d850c4a00eca5594356520ce158/scanner.py#L777
Isn't it better to just use `confirmation_retries`, which is what the user potentially defined as well?

**Review by Jomik — APPROVED** [review]
It looks good to me. Just had that one comment I would think about :)

File: scanner.py (hunk 5 above)
Comment (Jomik) [reply scanner.py:762]: We agreed to scale `max_acceptable_retries` off `confirmation_retries`. I added a scaling factor as a constant for now.

**Pull request merged**